HW4/HW4_SvTr.js

```diff
@@ -0,0 +1,224 @@
+    let key;
+
+    for (key in mainObj) {
+        objCopy[key] = mainObj[key]; // copy all properties into objCopy
+    }
+    return objCopy;
+}
+console.log(copy(menu));
+/*
+{0: {…}}
+0: {product_name: "burger menu", tax: 0.5, subtotal: 4.5, total: 5, items: Array(1), …}
+__proto__: Object
+ */
+
+
+// 2.2 Spread {... obj}
+var cloneMenu1 = { ...menu };
+console.log(cloneMenu1);
+/**
+ {0: {…}}
+0: {product_name: "burger menu", tax: 0.5, subtotal: 4.5, total: 5, items: Array(1), …}
+__proto__: Object
+ */
+
+
+//2.3 Object.assign({}, obj)
+const cloneMenu2 = Object.assign({}, menu);
+console.log(cloneMenu2);
+/*
+{0: {…}}
+0: {product_name: "burger menu", tax: 0.5, subtotal: 4.5, total: 5, items: Array(1), …}
+__proto__: Object
+*/
+
+//2.4 JSON cloning
+const cloneMenuJSON = JSON.parse(JSON.stringify(menu))//qwikly but
+//JSON cannot contain functions. This means that not every object can be cloned without loss.
+console.log(cloneMenuJSON);
+/*
+(1) [Object]
+0: Object
+produc t_name: "burger menu"
+tax: 0.5
+subtotal: 4.5
+total: 5
+items: Array(1)
+    0: Object
+    product_name: "french fries"
+    tax: 0
+    subtotal: 2
+    total: 2
+    entity_id: "someId2"
+    parent_id: "someId1"
+    items: Array(1)
+        0: Object
+        product_name: "ketchup"
+        tax: 0
+        subtotal: 1
+        total: 1
+        entity_id: "someId3"
+        parent_id: "someId2"
+        entity_id: "someId1"
+*/
+
+// 3. Create tree of nested objects
+// Use objects structure from example 2, generate flat Map set (key-value), here will be 
+//stored spesific user information.
+// for example(you created products collection with nested products inside),
+// by unic product id create list of product_id-product price.
+const menu3 = [{
+    product_name: "burger menu",
+    tax: 0.5,
+    subtotal: 4.5,
+    total: 5,
+    id: "someId1",
+    items: [{
+        product_name: "french fries",
+        tax: 0,
+        subtotal: 2,
+        total: 2,
+        id: "someId2",
+        items: [{
+            product_name: "ketchup",
+            tax: 0,
+            subtotal: 1,
+            total: 1,
+            id: "someId3"
+        }, {
+            product_name: "ketchupchik",
+            tax: 0,
+            subtotal: 1,
+            total: 1,
+            id: "someId3"
+        }, {
+            product_name: "ketchup",
+            tax: 0,
+            subtotal: 1,
+            total: 1,
+            id: "someId4"
+        }]
+    }]
+}];
+
+var menuItems = new Map();
+
+getMapNoDublicate(menu3);
+
+function getMapNoDublicate(data) {
+    for (let i = 0; i < data.length; i++) {
+        menuItems.set(data[i].id, data[i]);//method map.set() - add object if it does not exist (OR)
+        debugger;                          //                 - if it exist then updates element
+        //i can use iy like a checking
+        if (data[i].items && data[i].items.length) {//i check if element 'items' exist(true) and 
```

Comment:
in modern js you can use data[i].items?.length

Reply:
```
 if (data[i].items) {//i check if element 'items' exist(true) and 
            getMapNoDublicate(data[i].items);//recursively calls for nested 'items'
        }
```
if i change like that? What do you think?